### PR TITLE
added debruijn offset to ragg2 (-q)

### PR DESCRIPTION
**Pull request by four0four**
this patch follows up on a conversation with crowell in irc. ragg2 is now able to determine debruijn offsets as well as generate the patterns initially.

### CODE REVIEW

File: binr/ragg2/ragg2.c [hunk 1]
```diff
@@ -31,7 +31,8 @@ static int usage (int v) {
 	" -p [padding]    add padding after compilation (padding=n10s32)\n"
 	"                 ntas : begin nop, trap, 'a', sequence\n"
 	"                 NTAS : same as above, but at the end\n"
-	" -P [size]       prepend debrujn pattern\n"
+	" -P [size]       prepend debruijn pattern\n"
+	" -q [fragment]   debruijn pattern offset\n"
```
Comment (jvoisin) [inline binr/ragg2/ragg2.c]: Why `-q`?

**Comment by radare** [comment]
Update the manpage too.

And iirc this was alreay done in rafind

**Comment by crowell** [comment]
Nope this isnt in rafind

File: binr/ragg2/ragg2.c (hunk 1 above)
Comment (crowell) [inline binr/ragg2/ragg2.c]: @jvoisin `q`uery for offset? (also -o/O were already taken ;-))

**Comment by crowell** [comment]
@four0four update the manpage here https://github.com/radare/radare2/blob/master/man/ragg2.1

File: binr/ragg2/ragg2.c [hunk 2]
```diff
@@ -257,12 +260,16 @@ int main(int argc, char **argv) {
 		case 'z':
 			show_str = 1;
 			break;
+	  case 'q':
+			get_offset = 1;
+			sequence = strdup (optarg);
+      break;
```
Comment (crowell) [inline binr/ragg2/ragg2.c]: fix indenting

Comment (four0four) [inline binr/ragg2/ragg2.c]: can't believe I missed that. fixed.

**Comment by crowell** [comment]
LGTM

**Comment by crowell** [comment]
merged

**Comment by radare** [comment]
ok to merge, sorry, my memory failed :P and my lack of time didnt allowed me to review it faster

**Comment by alvarofe** [comment]
I'm getting the following. Am I using it well or I misunderstood?

```
$ ragg2 -P 138
4141414241414341414441414541414641414741414841414941414a41414b41414c41414d41414e41414f41415041415141415241415341415441415541415641415741415a41415941415a41416141416241416341416441416541416641416741416841416941416a41416b41416c41416d41416e41416f4141704141714141724141734141744141
~/projects/reverse/radare2/r2 @ (alvaro_fe) (master)
$ ragg2 -q 41415741415a4141
Little endian: -1
Big endian: -1
```

**Comment by four0four** [comment]
You do need to specify the 0x.. prefix (this is because the parsing for that argument involves dumping it into the math parser - I followed radare's implementation here). Could be changed pretty easily, I think. That said, this seems to be an issue with the code used to reverse those sequences - same thing occurs with woO:
```
[0x00404840]> woO 0x41434141
4
[0x00404840]> woO 0x4143414144414145
-1
```

There might be some assumption of 32 bit dwords somewhere - I don't know if this should be expected behavior.

**Comment by crowell** [comment]
Alvaro, add a -r to make it a string

**Comment by alvarofe** [comment]
```
$ ragg2 -P 128
4141414241414341414441414541414641414741414841414941414a41414b41414c41414d41414e41414f41415041415141415241415341415441415541415641415741415a41415941415a41416141416241416341416441416541416641416741416841416941416a41416b41416c41416d41416e41416f41417041417141
/tmp @ (alvaro_fe)
$ ragg2 -q 416c -r
Little endian: -1
Big endian: -1
```

**Comment by radare** [comment]
I think that both, hexpair and hex numbers should be supported here. The implementation is also asuming little endian?

**Comment by four0four** [comment]
I think you're right. We can check for the presence of -a (arch) for hints on endianness / word length - is there any code for that present in core already?

**Pull request closed without merging**